### PR TITLE
Update merge.ThreeWay() to allow very basic custom conflict resolution

### PR DESCRIPTION
**Pull request by cmasone-attic**
This patch modifies merge.ThreeWay() to take a callback that allows
for custom conflict resolution. The noms-merge command-line tool uses
this to inject a callback that accepts input from the console
dictating whether to accept the value from the 'left' or 'right' merge
candidates.

Toward #2445

### CODE REVIEW

**Comment by cmasone-attic** [comment]
Off I go to work on smush-merge :-)

You can review this now, if you want

File: samples/go/noms-merge/main.go [hunk 1]
```diff
@@ -91,6 +95,37 @@ func nomsMerge() error {
...
+	// TODO: Handle removes as well.
+	fmt.Printf("\nConflict at: %s\n", path.String())
+	fmt.Printf("Left:  %s\nRight: %s\n\n", left, right)
+	fmt.Println("Enter 'l' to accept the left value, 'r' to accept the right value")
+	var choice rune
+	for {
+		_, err := fmt.Scanf("%c", &choice)
+		d.PanicIfError(err)
+		switch choice {
+		case 'l', 'L':
```
Comment (aboodman) [inline samples/go/noms-merge/main.go]: Why don't I see this in red anywhere? The old version had it already.

Comment (cmasone-attic) [inline samples/go/noms-merge/main.go]: The old version definitely didn't have this code

**Comment by cmasone-attic** [comment]
alright! We now have 'mash'. It concatenates strings and then randomizes them. PTAL

**Comment by cmasone-attic** [comment]
@aboodman Do we want to review and land this? Should I ask @kalman to review since he's more familiar with this code?

**Comment by aboodman** [comment]
oh sorry, yeah. Seems reasonable to land.

**Comment by cmasone-attic** [comment]
You want to review it, or redirect to ben?

**Comment by aboodman** [comment]
ben please

File: samples/go/noms-merge/main.go [hunk 2]
```diff
@@ -91,6 +98,60 @@ func nomsMerge() error {
...
+			case types.String:
+				mashed := mash(a, b.(types.String))
+				fmt.Println("Replacing with", mashed)
+				return aChange, mashed, true
+			}
+		}
+	}
+}
+
+func mash(a, b types.String) types.String {
```
Comment (kalman) [inline samples/go/noms-merge/main.go]: wat?

File: samples/go/noms-merge/main.go [hunk 3]
```diff
@@ -91,6 +98,60 @@ func nomsMerge() error {
 	}))
 }
 
+func resolve(aChange, bChange types.ValueChanged, a, b types.Value, path types.Path) (change types.ValueChanged, merged types.Value, ok bool) {
```
Comment (kalman) [inline samples/go/noms-merge/main.go]: This needs to be tested. You may want to drive the majority of the merge logic outside of the stdio loop, like pass the choices over a channel.

File: go/merge/three_way.go [hunk 4]
```diff
@@ -11,6 +11,8 @@ import (
 	"github.com/attic-labs/noms/go/types"
 )
 
+type ResolveFunc func(aChange, bChange types.ValueChanged, a, b types.Value, path types.Path) (change types.ValueChanged, merged types.Value, ok bool)
```
Comment (kalman) [inline go/merge/three_way.go]: needs documentation, especially for me

**Comment by cmasone-attic** [comment]
ok, @kalman if you could look at this, I'd appreciate. The biggest change is that now there's a `merger` struct (which is a crap name) that represents a given merge attempt...sorta. There really isn't a lot of state to keep, other than the ValueReadWriter and the new conflict resolution callback. There will probably be more, though, so it seems like having a struct is The Right Thing.

The other big thing is that, rather than passing around `get`, `diff`, and (now) Path-concatenating closures for `a`, `b`, and `parent`, there's now a little interface that I can wrap Maps, Structs and Sets in.

File: go/merge/three_way.go [hunk 5]
```diff
@@ -200,7 +190,81 @@ func threeWayStructMerge(a, b, parent types.Struct, vwr types.ValueReadWriter, p
 		}
 		panic(fmt.Errorf("Bad key type in diff: %s", change.V.Type().Describe()))
 	}
-	return threeWayOrderedSequenceMerge(parent, aDiff, bDiff, makeGetFunc(a), makeGetFunc(b), makeGetFunc(parent), apply, vwr, progress)
+	return m.threeWayOrderedSequenceMerge(structCandidate{a}, structCandidate{b}, structCandidate{parent}, apply, path)
+}
+
+type candidate interface {
```
Comment (kalman) [inline go/merge/three_way.go]: I've spent some time trying to understand this change (with the mapCandidate etc). why did you need it?

**Comment by kalman** [comment]
let's start with these comments to help me get started.

File: samples/go/noms-merge/main.go (hunk 3 above)
Comment (cmasone-attic) [inline samples/go/noms-merge/main.go]: @aboodman, @rafael-atticlabs 
Are we really planning to keep this UX around? My understanding was that it was mostly for a demo, so I'm kinda loathe to refactor and add tests for it if we really don't care.

File: go/merge/three_way.go (hunk 5 above)
Comment (cmasone-attic) [inline go/merge/three_way.go]: I was explaining while you were reviewing!

The 'candidate' stuff was because the alternative was passing something like 7 closures to `threeWayOrderedSequenceMerge()`: one getFunc for each of a, b, and parent; one diffFunc for each of a and b; and one path-concatenation function for each of a and b.

That seemed absurd, so instead I made a little interface that Maps, Sets and Structs could all be adapted to and made the function take those instead. mapCandidate, setCandidate, and structCandidate are the adaptor implementations.

File: go/merge/three_way.go (hunk 4 above)
Comment (cmasone-attic) [inline go/merge/three_way.go]: // ResolveFunc is the type of a function which takes two conflicting changes and tries to resolve them. The two changes occured at `path`, one resulting in the Value `a` and the other in the Value `b`. If a resolution cannot be devised, `ok` should be false upon return and the other return values are undefined. If the conflict can be resolved, the function should return the appropriate types.ValueChanged to apply, the new value to be used when applying the change, and `true`.

(I will put this in the source too, obviously, but writing it here to hopefully explain)

File: go/merge/three_way.go (hunk 5 above)
Comment (kalman) [inline go/merge/three_way.go]: I see. Yes that would get pretty noisy, though I also find the adaptors quite distracting. What about another entirely different approach: if merger had a map of Kind --> getter/diff functions instead? (in fact it could be global)

File: samples/go/noms-merge/main.go (hunk 2 above)
Comment (cmasone-attic) [inline samples/go/noms-merge/main.go]: Aaron specifically asked for this to show that a conflict could be resolved using something other than the previously-existing value in one descendant or the other

File: samples/go/noms-merge/main.go [hunk 6]
```diff
@@ -91,6 +98,60 @@ func nomsMerge() error {
...
+			return aChange, a, true
+		case 'r', 'R':
+			return bChange, b, true
+		case 'm', 'M':
+			if !a.Type().Equals(b.Type()) {
+				fmt.Printf("Sorry, can't smush a %s with a %s\n", a.Type().Describe(), b.Type().Describe())
+			}
+			switch a := a.(type) {
+			case types.Bool:
+				return aChange, types.Bool(bool(a) || bool(b.(types.Bool))), true
```
Comment (kalman) [inline samples/go/noms-merge/main.go]: If I'm understanding correctly: is the use of aChange here even correct? If it's the change to apply, and we have a=false and b=true, shouldn't it return bChange?

Whether or not this code becomes relevant, i.e. if this stays around, it does raise the question whether the merge function should only return the change to apply, not the final value, because you can derive the final value by the change to apply. And, what if they disagree, as I mentioned with aChange?

Or alternatively don't return the change, only the final value, which would be simpler... but perhaps not enough information?

File: samples/go/noms-merge/main.go (hunk 2 above)
Comment (kalman) [inline samples/go/noms-merge/main.go]: Ok let's see what the decision is re keeping this code.

File: samples/go/noms-merge/main.go (hunk 6 above)
Comment (cmasone-attic) [inline samples/go/noms-merge/main.go]: The problem is that a `types.ValueChanged` doesn't contain the "changed value" for anything except a Set :-/ For maps, it's the key at which the described change happened, and for structs it's the field name. And, in those cases, the Value in the two ValueChanged structs is equal -- it has to be or we wouldn't be in conflict.

So, the value in aChange and bChange is either the same or it _should be_ ignored. I see now that it's not in the version of `apply()` used for Sets :-)

Comment (kalman) [inline samples/go/noms-merge/main.go]: aha, I see.

I think the proposed comment on the ResolveFunc documentation needs to be clearer then. And in fact since the value is always going to be the same in the ValueChanged struct, perhaps return a ChangeType instead?

File: go/merge/three_way.go (hunk 5 above)
Comment (cmasone-attic) [inline go/merge/three_way.go]: My gut reaction is that that will result is lots of type-switching, because `a`, `b`, and `parent` will come in as a types.Value. I feel like that's lame, because type-appropriate implementations of the same method is like...literally the thing you're supposed to use interfaces for.

What is it that's distracting? Is it that the definitions are all kind of in-line here? Is it the apply functions having to sorta un-candidate them? In my mind, the stuff that's complicated is all over in `threeWayOrderedSequenceMerge()`, and this use of adaptors doesn't change anything over there. This is really just shoehorning three different things into a similar enough shape that we can run them all through a single, common, complex code path.

File: samples/go/noms-merge/main.go (hunk 6 above)
Comment (cmasone-attic) [inline samples/go/noms-merge/main.go]: Can you propose a comment for ResolveFunc? The comment above _was_ my stab at explaining clearly and pithily and it apparently did not succeed

File: go/merge/three_way.go (hunk 5 above)
Comment (kalman) [inline go/merge/three_way.go]: > What is it that's distracting?

A few things, not necessarily in any order:
* PR diff is complicated
* figuring out whether these new collection types introduce any behavioral changes
* a fair amount of non-dense extra code
* boxing/unboxing is more code
* this entire algorithm is merging types.Collections and that's less obvious now
* I just find it less direct

I hear you on the type switching but I was more thinking like:

```go
differs := map[Kind]diffFunc{
  SetKind: func(v1, v2 types.Value, change chan<- types.ValueChanged, stop <-chan struct{}) types.Value {
    return v1.(Set).Diff(v2.(Set), change, stop)
  },
  MapKind: func(v1, v2 types.Value, change chan<- types.ValueChanged, stop <-chan struct{}) types.Value {
    return v1.(Map).Diff(v2.(Map), change, stop)
  },
  StructKind: func(v1, v2 types.Value, change chan<- types.ValueChanged, stop <-chan struct{}) types.Value {
    return v1.(Struct).Diff(v2.(Struct), change, stop)
  },
}
getters := map[Kind]getFunc{
  SetKind: func(v, k types.Value) types.Value {
    return k
  },
  MapKind: func(v, k types.Value) types.Value {
    return v.(Map).Get(k)
  },
  StructKind: func(v, k types.Value) types.Value {
    return v.(Struct).Get(string(k.(String)))
  },
}
pathConcat := map[Kind]pathConcatFunc { /* etc */ }
// then instead of aValue, bValue := a.get(aChange.V), b.get(bChange.V) have:
get := getters[a.Kind]
aValue, bValue := get(aChange.V), get(bChange.V)
```
now that I've typed that out I'm not sure it's any better. Is there a way to make the old callback approach more manageable? I can think of a few other ways to cut it.

File: samples/go/noms-merge/main.go (hunk 6 above)
Comment (kalman) [inline samples/go/noms-merge/main.go]: Like, make it clear that value is relative to the key that was changed. I read it as an entirely new value.

Also see change type? does that make sense? Because the key is going to be constant, seems moot.

Comment (cmasone-attic) [inline samples/go/noms-merge/main.go]: I've got a hopefully-better comment in my latest patch

File: samples/go/noms-merge/main.go (hunk 2 above)
Comment (cmasone-attic) [inline samples/go/noms-merge/main.go]: I've refactored this a little, passing in an io.Reader and io.Writer for i/o. That makes it testable without having to involve multiple goroutines and channels just for the sake of testing

**Comment by cmasone-attic** [comment]
Ok, PTAL. Lots of changes here. ResolveFunc is a bit different, the CLI resolve stuff has been tested, and I moved around the 'candidate' stuff to hopefully make the diff and the code more clear

File: go/merge/candidate.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+package merge
+
+import (
+	"fmt"
+
+	"github.com/attic-labs/noms/go/d"
+	"github.com/attic-labs/noms/go/types"
+)
+
+// candidate represents a collection that is a candidate to be merged. This interface exists to wrap Maps, Sets and Structs with a common API so that threeWayOrderedSequenceMerge() can remain agnostic to which kind of collections it's actually working with.
```
Comment (kalman) [inline go/merge/candidate.go]: Might as well wrap, even though this isn't public.

File: go/merge/candidate.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+type candidate interface {
```
Comment (kalman) [inline go/merge/candidate.go]: Thanks, this is much nicer.

File: go/merge/candidate.go [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+	diff(parent candidate, change chan<- types.ValueChanged, stop <-chan struct{})
+	get(k types.Value) types.Value
+	pathConcat(change types.ValueChanged, path types.Path) (out types.Path)
+	toValue() types.Value
```
Comment (kalman) [inline go/merge/candidate.go]: nit: getValue()?

File: go/merge/three_way.go [hunk 10]
```diff
@@ -11,7 +11,19 @@ import (
 	"github.com/attic-labs/noms/go/types"
 )
 
-// ErrMergeConflict indicates that a merge attempt failed and must be resolved manually for the provided reason.
+// ResolveFunc is the type for custom merge-conflict resolution callbacks.
+// When the merge algorithm encounters two non-mergeable changes (aChange and
+// bChange) at the same path, it calls the ResolveFunc passed into ThreeWay().
+// The callback gets the types of the two incompatible changes (added, changed
+// or removed) and the two Values that could not be merged (if any). If the
+// ResolveFunc cannot device a resolution, ok should be false upon return and
```
Comment (kalman) [inline go/merge/three_way.go]: cannot... _devise_?

File: go/merge/three_way.go [hunk 11]
```diff
@@ -11,7 +11,19 @@ import (
 	"github.com/attic-labs/noms/go/types"
 )
 
-// ErrMergeConflict indicates that a merge attempt failed and must be resolved manually for the provided reason.
+// ResolveFunc is the type for custom merge-conflict resolution callbacks.
+// When the merge algorithm encounters two non-mergeable changes (aChange and
+// bChange) at the same path, it calls the ResolveFunc passed into ThreeWay().
+// The callback gets the types of the two incompatible changes (added, changed
+// or removed) and the two Values that could not be merged (if any). If the
```
Comment (kalman) [inline go/merge/three_way.go]: Would it be correct to say "two keys" rather than "two Values"? Because this Value is different to the one referenced below "the new value to be used (if any)" and it would be helpful to disambiguate.

And if so - perhaps instead of saying "the new value to be used (if any)" say "the new value for that key (if any)"?

File: go/merge/three_way.go [hunk 12]
```diff
@@ -63,23 +82,26 @@ func unmergeable(a, b types.Value) bool {
 	return true
 }
 
-func updateProgress(progress chan struct{}) {
+type merger struct {
+	vwr      types.ValueReadWriter
```
Comment (kalman) [inline go/merge/three_way.go]: nit: vrw not vwr?

File: samples/go/noms-merge/main.go [hunk 13]
```diff
@@ -91,6 +99,52 @@ func nomsMerge() error {
...
+				return change, merged, false
+			}
+			switch a := a.(type) {
+			case types.Bool:
+				return aType, types.Bool(bool(a) || bool(b.(types.Bool))), true
+			case types.Number:
+				return aType, types.Number(float64(a) + float64(b.(types.Number))), true
+			case types.String:
+				concatenated := string(a) + string(b.(types.String))
+				fmt.Fprintln(out, "Replacing with", concatenated)
```
Comment (kalman) [inline samples/go/noms-merge/main.go]: Should this "replacing with" output be shown for bool/number as well?

**Comment by kalman** [comment]
lgtm

File: go/merge/candidate.go (hunk 7 above)
Comment (cmasone-attic) [inline go/merge/candidate.go]: done

File: go/merge/candidate.go (hunk 8 above)
Comment (cmasone-attic) [inline go/merge/candidate.go]: I'm glad it worked out :-)

File: go/merge/three_way.go (hunk 10 above)
Comment (cmasone-attic) [inline go/merge/three_way.go]: whoops!

File: go/merge/candidate.go (hunk 9 above)
Comment (cmasone-attic) [inline go/merge/candidate.go]: Yep, that's more accurate.

File: go/merge/three_way.go (hunk 11 above)
Comment (cmasone-attic) [inline go/merge/three_way.go]: But it's actually the two Values that couldn't be merged; if the code is trying to merge two maps {foo: bar} and {foo: baz}, `a` will be "bar" and `b` wil be "baz". And the thing that's being returned is the new value to use for 'foo', in the above example.

File: go/merge/three_way.go (hunk 12 above)
Comment (cmasone-attic) [inline go/merge/three_way.go]: ...I may have made this mistake in many, many places. Accursèd mild dyslexia!

File: samples/go/noms-merge/main.go (hunk 13 above)
Comment (cmasone-attic) [inline samples/go/noms-merge/main.go]: Done

File: go/merge/three_way.go (hunk 11 above)
Comment (kalman) [inline go/merge/three_way.go]: Ah, got it. Never mind then.
